[espnow] Small changes and fixes

nielsnl68 opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->
This PR resolved some small issues found in PR #9582


## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5207

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

nielsnl68 committed:
remove debug print

nielsnl68 committed:
move receiving data in to a Task

nielsnl68 committed:
fix enabke() method

nielsnl68 committed:
delete only the peers from espnow self not from the component as well

nielsnl68 committed:
adding a fix

nielsnl68 committed:
reorder some is if statements

jesserockz commented on esphome/components/espnow/automation.h:
```diff
@@ -40,20 +40,20 @@ template<typename... Ts> class SendAction : public Action<Ts...>, public Parente
-        } else if (!this->sent_.empty()) {
+        if (!this->sent_.empty()) {
           this->sent_.play(x...);
+        } else if (this->sent_.empty() && this->flags_.wait_for_sent) {
```
```suggestion
        } else if (this->flags_.wait_for_sent) {
```

jesserockz commented on esphome/components/espnow/espnow_component.cpp:
```diff
@@ -213,8 +293,13 @@ void ESPNowComponent::enable_() {
 
   for (auto peer : this->peers_) {
-    this->add_peer(peer.address);
+    if (esp_now_is_peer_exist(peer.address)) {
```
What is the reason for this change? 
When deinit it's called, according to the documentation, all peers are removed, so the list will be empty on init

nielsnl68 replied:
oeps my mistake. this is the wrong place. this should happen in the disable() method.

jesserockz reviewed (changes requested):
Putting the entire code into a task is bad. We now will have a task potentially putting items into the peer list, if an add peer action was called at the same time, it would corrupt and crash.

Also calling the handlers from the task is bad, now we would have code external to this component running without supervision, also potentially causing corruption. 

I can't agree with this change unless there is significant restructuring to ensure cross thread safety. 

The other little fixes are ok and can be merged if the bigger change is removed

nielsnl68 committed:
- rework how "`on_unknown_peer`"  is handled. WIP
- get correct wifi channel
- fix removing peers at the wrong moment.
- add get_wifi_channel()

nielsnl68 committed:
- add get_wifi_channel()

nielsnl68 committed:
Merge branch 'dev' into nvds-espnow-fixes-020825

nielsnl68 committed:
Merge branch 'dev' into nvds-espnow-fixes-020825

nielsnl68 committed:
Remove task and move the code back in the loop()

nielsnl68 committed:
Merge branch 'nvds-espnow-fixes-020825' of https://github.com/nielsnl68/esphome into nvds-espnow-fixes-020825

nielsnl68 committed:
Update esphome/components/espnow/automation.h

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nielsnl68 committed:
Merge branch 'dev' into nvds-espnow-fixes-020825

jesserockz commented on esphome/components/espnow/espnow_component.cpp:
```diff
@@ -229,7 +227,9 @@ void ESPNowComponent::disable() {
-    this->del_peer(peer.address);
+    if (esp_now_is_peer_exist(peer.address)) {
+      esp_now_del_peer(peer.address);
+    }
```
I dont think this entire block is even needed:

> When `esp_now_deinit()` is called, all of the information of paired devices are deleted.

https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/network/esp_now.html#initialization-and-deinitialization

jesserockz commented on esphome/components/espnow/espnow_component.cpp:
```diff
@@ -178,11 +178,8 @@ void ESPNowComponent::enable_() {
-  }
-#endif
+  wifi_second_chan_t dummy;
+  esp_wifi_get_channel(&this->wifi_channel_, &dummy);
```
```suggestion
  this->get_wifi_channel();
```

nielsnl68 committed:
Update esphome/components/espnow/espnow_component.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nielsnl68 commented on esphome/components/espnow/espnow_component.cpp:
```diff
@@ -229,7 +226,9 @@ void ESPNowComponent::disable() {
   esp_now_unregister_send_cb();
 
   for (auto peer : this->peers_) {
-    this->del_peer(peer.address);
+    if (esp_now_is_peer_exist(peer.address)) {
+      esp_now_del_peer(peer.address);
+    }
   }
```
```suggestion
```

nielsnl68 committed:
remove un needed peer clearing

nielsnl68 committed:
Merge branch 'dev' into nvds-espnow-fixes-020825

nielsnl68 committed:
Merge branch 'dev' into nvds-espnow-fixes-020825
